Implements a PID controller by composing a Diagram

amcastro-tri opened the pull request:
This PR closes #3215. It implements a PID controller by composing smaller primitives into a Diagram.

cc'ing @naveenoid because you were interested on this.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3275)
<!-- Reviewable:end -->

amcastro-tri commented:
The usual suspects: @sherm1, @david-german-tri 

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQ7uPhi6VaBD7NBbeqa:bqpa561)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
The usual suspects: +@sherm1, +@david-german-tri

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQ7uVoOa25WtSu9SnuX:b-8jo617)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Starting review now.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQBiJgEqRFlQbA19l57:b-brdniz)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
This is super exciting!

---

Reviewed 4 of 10 files at r1.
Review status: 4 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/framework/diagram.h, line 257 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBj_FiE8yUMaYSE19V:-KQBiuSNFN9v0EyuUVYa:b-mpdqrq) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L257)):*
> ```objc
>   /// Classes inheriting from %Diagram need access to this method in order to
>   /// pass their constituyent sub-system's the apropriate sub-context.
>   ContextBase<T>* GetMutableSubSystemContext(
> ```

I'm not sure this accessor should exist.  A mutable pointer to the whole subcontext makes it possible to mess with the input and output, but for your usage below you only need the state.  How about just using `GetMutableSubsystemState` instead?

---

*[drake/systems/framework/diagram_context.h, line 216 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkQElexsICGLNq0Nh:-KQBkQElexsICGLNq0Ni:bs4q9op) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram_context.h#L216)):*
> ```objc
>   /// Returns a mutable context structure for a given constituent system @p sys,
>   /// or nullptr if @p sys is not a constituent system.
>   ContextBase<T>* GetMutableSubsystemContext(SystemIndex sys) {
> ```

Ditto.

---

*[drake/systems/framework/primitives/integrator.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkTAdpq6Z_EOFhrhj:-KQBkTAdpq6Z_EOFhrhk:buwm3bs) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L26)):*
> ```cpp
> template <typename T>
> void Integrator<T>::set_integral_value(
>     ContextBase<T>* context, const Eigen::Ref<const VectorX<T>>& value) const {
> ```

Why not just `ContinuousState<T>* state` instead of `context`?

---

*[drake/systems/framework/primitives/integrator.cc, line 27 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkXWc_AAbCi6WxSFv:-KQBkXWc_AAbCi6WxSFw:bvd0miy) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L27)):*
> ```cpp
> void Integrator<T>::set_integral_value(
>     ContextBase<T>* context, const Eigen::Ref<const VectorX<T>>& value) const {
>   // TODO(amcastro-tri): Provide simple accessors here to avoid lengthy
> ```

BTW, I totally agree, digging out state vectors should be sugared.  But passing `ContinuousState<T>` to this function would also help!

---

*[drake/systems/framework/primitives/pid_controller.cc, line 1 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBlY1g0kMjvvEa96AZ:-KQBlY1g0kMjvvEa96A_:b-3gahxg) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.cc#L1)):*
> ```cpp
> #include "drake/systems/framework/primitives/pid_controller.h"
> ```

BTW, I assume you intentionally didn't use `-inl.h`?

---

*[drake/systems/framework/primitives/test/pid_controller_test.cc, line 53 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBlj3vsP6v-_i7tAY8:-KQBlj3vsP6v-_i7tAY9:b53lyvy) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/test/pid_controller_test.cc#L53)):*
> ```cpp
>   context->SetInputPort(1, MakeInput(std::move(vec1)));
> 
>   // Initializes the controllers' context to the default value in which the
> ```

This should be three separate `TEST_F` cases, with the common initialization in `SetUp`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQBlqEj-ksnrTUT6FG7:bfl06os)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 4 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/diagram.h, line 257 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBj_FiE8yUMaYSE19V:-KQD5QkJJKYLVR1bWjQ5:b-92jbrh) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L257)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I'm not sure this accessor should exist.  A mutable pointer to the whole subcontext makes it possible to mess with the input and output, but for your usage below you only need the state.  How about just using `GetMutableSubsystemState` instead?
</details>
Never mind, after further poking around today, I think this is clearly needed for Diagrams that contain Diagrams.

---

*[drake/systems/framework/diagram_context.h, line 216 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkQElexsICGLNq0Nh:-KQD5_HqSpKYDbzseA5k:b-vlwd2b) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram_context.h#L216)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Ditto.
</details>
Never mind, see above.

---

*[drake/systems/framework/primitives/integrator.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkTAdpq6Z_EOFhrhj:-KQD5bPCBJMb4sYvN7MP:b-zbfic5) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L26)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Why not just `ContinuousState<T>* state` instead of `context`?
</details>
I do still think this question stands, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Just a few answers.

---

Review status: 4 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/diagram.h, line 257 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBj_FiE8yUMaYSE19V:-KQGTV9M5pRLuQXWULXR:b-7oobdr) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L257)):*
> A mutable pointer to the whole subcontext makes it possible to mess with the input and output,

I thought the output was not in the context? did that change?

---

*[drake/systems/framework/primitives/integrator.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkTAdpq6Z_EOFhrhj:-KQGS_QxguU58hL4cCjS:bvcb9j) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L26)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I do still think this question stands, though.
</details>
Good point. But then as we add more an more primitives I think it is easier for users to just pass context's around and have a homogeneous API for all systems (continuous, discrete, hybrid). Then the specific primitive know what to dig out of it. 

---

*[drake/systems/framework/primitives/pid_controller.cc, line 1 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBlY1g0kMjvvEa96AZ:-KQGT2QIUnObJAeYZUmL:b-6qn5ox) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.cc#L1)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, I assume you intentionally didn't use `-inl.h`?
</details>
Yes. I was wondering how the proposed method in #3202 would change things and decided to wait.

---

*[drake/systems/framework/primitives/test/pid_controller_test.cc, line 53 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBlj3vsP6v-_i7tAY8:-KQGTgcO8hqz9LRJVjku:bm363nk) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/test/pid_controller_test.cc#L53)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> This should be three separate `TEST_F` cases, with the common initialization in `SetUp`.
</details>
I agree. I will split these tests once we are done with the spring. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQGTdUnRiS8d42d20-3:b-3xf51t)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Part way through. This is awesome! PTAL.

---

Reviewed 6 of 10 files at r1.
Review status: 7 of 10 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/diagram.h, line 254 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJLnlFmh4xSkbG0cLZ:-KQJLnlG_VHOenF0gfTW:b-s15rm5) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L254)):*
> ```objc
>   Diagram() {}
> 
>   /// Returns the sub-context that corresponds to the system @p sub_system.
> ```

BTW, can we agree just to call these "subcontext" and "subsystem" without the extra punctuation?

---

*[drake/systems/framework/diagram.h, line 256 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJLxN8617qyLv8HrtX:-KQJLxN8617qyLv8HrtY:bmszbux) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L256)):*
> ```objc
>   /// Returns the sub-context that corresponds to the system @p sub_system.
>   /// Classes inheriting from %Diagram need access to this method in order to
>   /// pass their constituyent sub-system's the apropriate sub-context.
> ```

several typos and extra punctuation: "pass their constituent subsystems the appropriate subcontext."

---

*[drake/systems/framework/diagram.h, line 257 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJMOIUsoMjwjc9MPde:-KQJMOIUsoMjwjc9MPdf:bx7u1us) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L257)):*
> ```objc
>   /// Classes inheriting from %Diagram need access to this method in order to
>   /// pass their constituyent sub-system's the apropriate sub-context.
>   ContextBase<T>* GetMutableSubSystemContext(
> ```

Call this "GetMutableSubsystemContext" without the extra cap in "SubSystem".

---

*[drake/systems/framework/diagram_context.h, line 217 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJMeAw7l7F1IcJCzDq:-KQJMeAw7l7F1IcJCzDr:b1s9i36) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram_context.h#L217)):*
> ```objc
>   /// or nullptr if @p sys is not a constituent system.
>   ContextBase<T>* GetMutableSubsystemContext(SystemIndex sys) {
>     auto it = contexts_.find(sys);
> ```

Yikes! Why does this require a search rather than an indexing operation? Isn't this on the critical path to getting a value out of the context?

---

*[drake/systems/framework/primitives/integrator.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkTAdpq6Z_EOFhrhj:-KQJN2Qg6olyq6ZDm18E:be76dl9) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L26)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Good point. But then as we add more an more primitives I think it is easier for users to just pass context's around and have a homogeneous API for all systems (continuous, discrete, hybrid). Then the specific primitive know what to dig out of it. 
</details>
BTW, this also came up in #3202. There is a lot to be said in mutable-context methods for supplying only the piece that can be legitimately modified (here just the Integrator's state). However, I also see @amcastro-tri's point about a uniform API -- trying to make this bulletproof might result in an awkward API. If we do go the bulletproof route, the generic signature still requires a context. It would look something like this:
```c++
void SomeMethod(const Context& context, MutableContextThing* thing, ...);
```
where "mutable context thing" would refer to something in the same `context` but would prevent writing on anything else. (The full context is necessary for read-only access to data that is required for calculating the mutable thing.)

I have mixed feelings about this. In Simbody methods always take either a whole const Context or a whole mutable Context. That didn't cause any trouble in practice but I do think we should consider whether to go fine grained in our API. One advantage is the method code doesn't have to know how to dig out the right piece of the context (as is being done below). Maybe file an issue and not deal with this in this PR?

---

*[drake/systems/framework/primitives/pid_controller.cc, line 48 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJPXems3jz3sBny6m5:-KQJPXems3jz3sBny6m6:b-tifqrz) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.cc#L48)):*
> ```cpp
> template <typename T>
> bool PidController<T>::has_any_direct_feedthrough() const {
>   return true;
> ```

BTW, maybe add TODO to make this depend on Kp != 0? Or is that overkill.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQJPqRgK_yar_unDg2G:b-rvaikx)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Finished review pass.

---

Reviewed 4 of 10 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/pid_controller.h, line 15 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOPpJZTWBf8mWVzT0n:-KQOPpJZTWBf8mWVzT0o:b-pyit9p) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L15)):*
> ```objc
> namespace systems {
> 
> /// A PID controller system. Given an error signal `e` and its derivative `edot`
> ```

"its derivative" -> "its *time* derivative"

---

*[drake/systems/framework/primitives/pid_controller.h, line 18 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOQ0ACcFLdvPu2tNlg:-KQOQ0ACcFLdvPu2tNlh:b-g1pd11) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L18)):*
> ```objc
> /// the output of this sytem is
> /// \f[
> ///   y = k_p \varepsilon + k_i \int{\varepsilon~dt} + k_d \dot{\varepsilon}
> ```

This expression is inconsistent with the comment that used `e` rather than epsilon. Also, using this makes it much harder for someone reading the header file to discern the equation (they have to be a human Latex parser). What about
```
  y = kp e + ki integ(e,dt) + kd edot
  where integ(e,dt) is the time integral of e
```
That would be readable here and equally readable in doxygen, though not as beautiful there.

---

*[drake/systems/framework/primitives/pid_controller.h, line 42 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQORa9D7JN8tQFCvHjY:-KQORa9D7JN8tQFCvHjZ:bnylsdp) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L42)):*
> ```objc
> 
>   /// Sets @p context to a default state in which the integral of the
>   /// controller is zero.
> ```

BTW, "integral of the error signal is zero"?

---

*[drake/systems/framework/primitives/pid_controller.h, line 45 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQORiHkMTfRYlMHlF8s:-KQORiHkMTfRYlMHlF8t:b-kt62im) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L45)):*
> ```objc
>   void SetDefaultState(ContextBase<T>* context) const;
> 
>   /// Sets the integral of the %PidController to zero.
> ```

The above comment is wrong.
BTW, do you need both `SetDefaultState()` and `set_integral_value()`?

---

*[drake/systems/framework/primitives/pid_controller.h, line 54 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOPKRW4osxoEgWnB8S:-KQOPKRW4osxoEgWnB8T:b-qn22j4) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L54)):*
> ```objc
> 
>   /// Returns the input port to the time derivative or rate of the error signal.
>   const SystemPortDescriptor<T>& get_error_signal_rate_port() const;
> ```

This name is odd and I'm not sure "signal" helps. How about calling the two methods `get_error_port()` and `get_error_derivative_port()` (matches the "D" in PID), or `get_error_rate_port()`?

---

*[drake/systems/framework/primitives/test/pid_controller_test.cc, line 41 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOSh2EzXJzvCx2KJMV:-KQOSh2EzXJzvCx2KJMW:bcvp3o6) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/test/pid_controller_test.cc#L41)):*
> ```cpp
>   // Creates a "fake" free-standing input port.
>   Vector3<double> error_signal(1.0, 2.0, 3.0);
>   Vector3<double> error_signal_rate(1.3, 0.9, 3.14);
> ```

BTW, `error_rate` or `error_rate_signal` would be better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQOT9vq5inWlhgakx_y:b-drlv36)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 6 of 12 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/diagram.h, line 254 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJLnlFmh4xSkbG0cLZ:-KQQL5gHAnxIVwLAbeMt:b9uvjk6) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L254)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, can we agree just to call these "subcontext" and "subsystem" without the extra punctuation?
</details>
sure. Updated doc and method name.

---

*[drake/systems/framework/diagram.h, line 256 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJLxN8617qyLv8HrtX:-KQQLbIimKcmeBWQcszd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L256)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> several typos and extra punctuation: "pass their constituent subsystems the appropriate subcontext."
</details>
Done.

---

*[drake/systems/framework/diagram.h, line 257 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJMOIUsoMjwjc9MPde:-KQQLdOqD5-TMkFVdFJ2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L257)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Call this "GetMutableSubsystemContext" without the extra cap in "SubSystem".
</details>
Done.

---

*[drake/systems/framework/diagram_context.h, line 217 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJMeAw7l7F1IcJCzDq:-KQQLxAfhj7PnuhuSMNd:bfkzmt4) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram_context.h#L217)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Yikes! Why does this require a search rather than an indexing operation? Isn't this on the critical path to getting a value out of the context?
</details>
I agree. One solution that I proposed is at `Finalize()` stage to number systems and to have a member `System<T>::sys_index_` (that'd be the simple but efficient Fortran solution :-)) but I believe @david-german-tri had a more elegant solution in mind equally efficient.  I also believe this is the best we can do with System 2.0 as it is in its current version.

---

*[drake/systems/framework/primitives/integrator.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkTAdpq6Z_EOFhrhj:-KQQNHOiTFt5wZlsXBfb:bql2d3w) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L26)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, this also came up in #3202. There is a lot to be said in mutable-context methods for supplying only the piece that can be legitimately modified (here just the Integrator's state). However, I also see @amcastro-tri's point about a uniform API -- trying to make this bulletproof might result in an awkward API. If we do go the bulletproof route, the generic signature still requires a context. It would look something like this:
> ```c++
> void SomeMethod(const Context& context, MutableContextThing* thing, ...);
> ```
> where "mutable context thing" would refer to something in the same `context` but would prevent writing on anything else. (The full context is necessary for read-only access to data that is required for calculating the mutable thing.)
> 
> I have mixed feelings about this. In Simbody methods always take either a whole const Context or a whole mutable Context. That didn't cause any trouble in practice but I do think we should consider whether to go fine grained in our API. One advantage is the method code doesn't have to know how to dig out the right piece of the context (as is being done below). Maybe file an issue and not deal with this in this PR?
</details>
Knowing there is a "continuous state" (or anything else) in the context seems like it'd be overwhelming for newbie users of System 2.0. In the case of this method, having to set the integral value is something newbie users (not having any idea at all of what the context my have in it) will do all the time and most of times they **must** do it to solve their particular problem.
I opt for leaving this one signature as it is. The implementer of `Integrator` (which would know a bit better than a newbie) should know what to dig out since is the one who requested resources in the context in the first place.
I am open to suggestions though. I like this example because as simple as it is it shows the kind of API "problems" that we need to solve.  

---

*[drake/systems/framework/primitives/integrator.cc, line 27 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkXWc_AAbCi6WxSFv:-KQQONXC0g1g6FiplFh1:br98617) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L27)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, I totally agree, digging out state vectors should be sugared.  But passing `ContinuousState<T>` to this function would also help!
</details>
see my previous reply.

---

*[drake/systems/framework/primitives/pid_controller.cc, line 48 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJPXems3jz3sBny6m5:-KQQSOCWgAobjPMdpqx0:b-yfxjyf) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.cc#L48)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, maybe add TODO to make this depend on Kp != 0? Or is that overkill.
</details>
It is not. I was just sprinting. Done.

---

*[drake/systems/framework/primitives/pid_controller.h, line 15 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOPpJZTWBf8mWVzT0n:-KQQedMkK4NkTKrUH7b0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L15)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> "its derivative" -> "its *time* derivative"
</details>
Done.

---

*[drake/systems/framework/primitives/pid_controller.h, line 18 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOQ0ACcFLdvPu2tNlg:-KQQfLXidAwPY7OZlrTB:b-7m6dox) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L18)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This expression is inconsistent with the comment that used `e` rather than epsilon. Also, using this makes it much harder for someone reading the header file to discern the equation (they have to be a human Latex parser). What about
> ```
>   y = kp e + ki integ(e,dt) + kd edot
>   where integ(e,dt) is the time integral of e
> ```
> That would be readable here and equally readable in doxygen, though not as beautiful there.
</details>
I changed to the text version. So part of our "style" is not to use Latex even when Doxygen supports it?

---

*[drake/systems/framework/primitives/pid_controller.h, line 42 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQORa9D7JN8tQFCvHjY:-KQQhzkIDAi6wiYXGmPa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L42)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, "integral of the error signal is zero"?
</details>
Done.

---

*[drake/systems/framework/primitives/pid_controller.h, line 45 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQORiHkMTfRYlMHlF8s:-KQQiFqg0vcnO6rc2JIh:b-b0vq9d) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L45)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> The above comment is wrong.
> BTW, do you need both `SetDefaultState()` and `set_integral_value()`?
</details>
Done. I think having both is just convenient.

---

*[drake/systems/framework/primitives/pid_controller.h, line 54 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOPKRW4osxoEgWnB8S:-KQQitHBzoUdgwFd3-5W:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L54)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This name is odd and I'm not sure "signal" helps. How about calling the two methods `get_error_port()` and `get_error_derivative_port()` (matches the "D" in PID), or `get_error_rate_port()`?
</details>
Done.

---

*[drake/systems/framework/primitives/test/pid_controller_test.cc, line 41 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOSh2EzXJzvCx2KJMV:-KQQjDMao_SUm4saZrhF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/test/pid_controller_test.cc#L41)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, `error_rate` or `error_rate_signal` would be better.
</details>
Done.

---

*[drake/systems/framework/primitives/test/pid_controller_test.cc, line 53 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBlj3vsP6v-_i7tAY8:-KQQox40KaLZwZv8rfi9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/test/pid_controller_test.cc#L53)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I agree. I will split these tests once we are done with the spring. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 6 files at r2.
Review status: 8 of 12 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/diagram.h, line 254 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJLnlFmh4xSkbG0cLZ:-KQBj2riG1ZJQvBNHcaV:b-8ztwc3) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L254)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> sure. Updated doc and method name.
</details>
LGTM, and +1 for this spelling.

---

*[drake/systems/framework/diagram.h, line 257 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBj_FiE8yUMaYSE19V:-KQQpg00yPtm5LGC0IT2:b-jkawxz) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L257)):*
> I thought the output was not in the context? did that change?

You're right, I misspoke.  Also, this accessor now exists on master, which at least partially explains the conflicts that GitHub is reporting.

---

*[drake/systems/framework/diagram_context.h, line 217 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJMeAw7l7F1IcJCzDq:-KQQq_OUTUA3jxyY38oF:b-k415xo) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram_context.h#L217)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I agree. One solution that I proposed is at `Finalize()` stage to number systems and to have a member `System<T>::sys_index_` (that'd be the simple but efficient Fortran solution :-)) but I believe @david-german-tri had a more elegant solution in mind equally efficient.  I also believe this is the best we can do with System 2.0 as it is in its current version.
</details>
Yeah, we can just make this a vector instead of a map, with no API impact.  I have some draft changes in progress.

---

*[drake/systems/framework/primitives/integrator.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkTAdpq6Z_EOFhrhj:-KQQsKb-9TQweqdKwvwU:b-2uwoz4) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L26)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Knowing there is a "continuous state" (or anything else) in the context seems like it'd be overwhelming for newbie users of System 2.0. In the case of this method, having to set the integral value is something newbie users (not having any idea at all of what the context my have in it) will do all the time and most of times they **must** do it to solve their particular problem.
> I opt for leaving this one signature as it is. The implementer of `Integrator` (which would know a bit better than a newbie) should know what to dig out since is the one who requested resources in the context in the first place.
> I am open to suggestions though. I like this example because as simple as it is it shows the kind of API "problems" that we need to solve.  
</details>
I definitely agree this is fine as-is for the current PR.  Opened #3308 for discussion.

---

*[drake/systems/framework/primitives/pid_controller.cc, line 48 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJPXems3jz3sBny6m5:-KQQtggb-LxaDBUDhD91:bm9ao1a) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.cc#L48)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> It is not. I was just sprinting. Done.
</details>
Oh, that's nice!  I mean, it's a bit of a hack for general sparsity, but it's a nice hack.

---

*[drake/systems/framework/primitives/pid_controller.h, line 18 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOQ0ACcFLdvPu2tNlg:-KQQu0wSdqCvmYlmlDrG:b-wuppn9) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L18)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I changed to the text version. So part of our "style" is not to use Latex even when Doxygen supports it?
</details>
FWIW, I think it's perfectly fine to expect readers to be human LaTeX parsers, but I would argue against a style rule on the subject, pro or con. Also, TIL doxygen supports LaTeX.

---

*[drake/systems/framework/primitives/pid_controller.h, line 45 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQORiHkMTfRYlMHlF8s:-KQQuc8y1LF0lFIjDjJe:b-5hgkxl) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L45)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Done. I think having both is just convenient.
</details>
At times, it's been suggested that `SetDefaultState` should be a pure virtual function on `System<T>`.  Let's not rush to do that, but I think this PR bolsters the argument for it a little.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQQuyduxKMR8M4aDOd9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 8 of 12 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/diagram_context.h, line 217 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJMeAw7l7F1IcJCzDq:-KQR3LGByrzq3LMCNDfW:b-x31bx7) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram_context.h#L217)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Yeah, we can just make this a vector instead of a map, with no API impact.  I have some draft changes in progress.
</details>
Yaiii!

---

*[drake/systems/framework/primitives/integrator.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBkTAdpq6Z_EOFhrhj:-KQR3WIovYX_0kXd-EF5:b-243kym) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/integrator.cc#L26)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I definitely agree this is fine as-is for the current PR.  Opened #3308 for discussion.
</details>
Great! thank you

---

*[drake/systems/framework/primitives/pid_controller.cc, line 48 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJPXems3jz3sBny6m5:-KQR3eEuAgcfiRhvP9Ob:b-w5yuer) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.cc#L48)):*
> it's a bit of a hack for general sparsity

why? it should describe the sparsity correctly right?

---

*[drake/systems/framework/primitives/pid_controller.h, line 18 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOQ0ACcFLdvPu2tNlg:-KQR3qCBdrgNgzsNe5MH:b-xke7uh) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L18)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> FWIW, I think it's perfectly fine to expect readers to be human LaTeX parsers, but I would argue against a style rule on the subject, pro or con. Also, TIL doxygen supports LaTeX.
</details>
Yes! and it looks beautiful!

---

*[drake/systems/framework/primitives/pid_controller.h, line 45 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQORiHkMTfRYlMHlF8s:-KQR3w5Ir3j7H6sYNn-c:b7ins6v) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L45)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> At times, it's been suggested that `SetDefaultState` should be a pure virtual function on `System<T>`.  Let's not rush to do that, but I think this PR bolsters the argument for it a little.
</details>
ah! I actually was thinking about adding `SetDefaultState` to `System` in this PR. However I decided to wait until I had more usage examples (like the kuka demo :-) )

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 8 of 12 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/pid_controller.cc, line 48 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJPXems3jz3sBny6m5:-KQR4zySPuOqxIEmAndZ:b-91zksr) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.cc#L48)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> > it's a bit of a hack for general sparsity
> 
> why? it should describe the sparsity correctly right?
> 
</details>
Yes, absolutely; I just mean we should eventually have a more general notation for sparsity as a function of input port and output port.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 5 of 11 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/systems/framework/diagram.h, line 257 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQBj_FiE8yUMaYSE19V:-KQR71BElWT32giwIgFe:b-xvxlj2) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/diagram.h#L257)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> > I thought the output was not in the context? did that change?
> 
> You're right, I misspoke.  Also, this accessor now exists on master, which at least partially explains the conflicts that GitHub is reporting.
</details>
Rebased  on master which yes, already had this accessor.

---

*[drake/systems/framework/primitives/pid_controller.cc, line 48 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQJPXems3jz3sBny6m5:-KQR6rUr-v3Lb1AlilwL:b-f2a2ow) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.cc#L48)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Yes, absolutely; I just mean we should eventually have a more general notation for sparsity as a function of input port and output port.
</details>
ah! yes! completely agree with that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review pass done. A few PTAL comments.

---

Reviewed 4 of 6 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/gain.h, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQRny0M0VreH9G2hISs:-KQRny0M0VreH9G2hISt:bvz7ipo) ([raw file](https://github.com/robotlocomotion/drake/blob/1ed01c073c4646c1e80457eb877becb75f9c7999/drake/systems/framework/primitives/gain.h#L35)):*
> ```objc
> 
>   /// Returns the gain constant.
>   T get_gain() const;
> ```

BTW, `const T&` in case this is a big AutoDiffScalar?

---

*[drake/systems/framework/primitives/pid_controller.cc, line 63 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQRj8Mpu1-Xz8btwTsO:-KQRj8Mpu1-Xz8btwTsP:bh3dabc) ([raw file](https://github.com/robotlocomotion/drake/blob/1ed01c073c4646c1e80457eb877becb75f9c7999/drake/systems/framework/primitives/pid_controller.cc#L63)):*
> ```cpp
> template <typename T>
> bool PidController<T>::has_any_direct_feedthrough() const {
>   if (get_Kp() == 0 && get_Kd() == 0) return false;
> ```

BTW, this could cause trouble if Kp and Kd can be changed after construction since there is no mechanism to note a change in feedthrough. When we have Parameters they can note changes and could force re-Finalize() of the Diagram. @david-german-tri, does the DiagramBuilder idea preclude this? 

---

*[drake/systems/framework/primitives/pid_controller.h, line 18 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQOQ0ACcFLdvPu2tNlg:-KQRkb1SCVB2W5MPdCos:bqvdklo) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L18)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Yes! and it looks beautiful!
</details>
I think forced Latex parsing is too much to ask of humans. I for one can't do it except in simple cases. It is sad to give up the formatted beauty but it is also sad to have header files that can't be interpreted without generating the doxygen.

When the equations aren't clear in the comments, it is very hard to verify that the code matches. For mathematical code it is best to have comments presenting the math immediately followed by code that allegedly implements those equations. Since the code is limited to ASCII text, it is best for the equations to be written that way also to facilitate direct comparison.

---

*[drake/systems/framework/primitives/pid_controller.h, line 45 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQORiHkMTfRYlMHlF8s:-KQRlgY4wLEMnx2EaLo6:ben9gqc) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L45)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> ah! I actually was thinking about adding `SetDefaultState` to `System` in this PR. However I decided to wait until I had more usage examples (like the kuka demo :-) )
</details>
BTW, is it clear what you would want this to do? Would `SetDefaultState()` restore a Context to the condition it was in right after `CreateDefaultContext()`? Or affect only continuous states? Other states are: difference states xd, modes, input and output port samples. Time and parameter values might also be included. Is this a proposal for a new family of System methods?

---

*[drake/systems/framework/primitives/pid_controller.h, line 22 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQRkNAnKCwo0J8ZJuEw:-KQRkNAnKCwo0J8ZJuEx:b-rs7926) ([raw file](https://github.com/robotlocomotion/drake/blob/1ed01c073c4646c1e80457eb877becb75f9c7999/drake/systems/framework/primitives/pid_controller.h#L22)):*
> ```objc
> /// where `integ(e,dt)` is the time integral of `e`.
> /// When the proportional constant is non-zero the input to this system
> /// directly feeds through to its output.
> ```

This comment needs updating since the same thing happens with the derivative constant (for a different input port).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQRoF-ur1meLithWyCQ:b-5fnc6o)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@drake-jenkins-bot windows-msvc-ninja-32-experimental-debug please

amcastro-tri commented:
@drake-jenkins-bot windows-msvc-ninja-32-experimental please

amcastro-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/gain.h, line 35 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQRny0M0VreH9G2hISs:-KQRyV05dcKTDfFLCVIk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ed01c073c4646c1e80457eb877becb75f9c7999/drake/systems/framework/primitives/gain.h#L35)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, `const T&` in case this is a big AutoDiffScalar?
</details>
Done.

---

*[drake/systems/framework/primitives/pid_controller.h, line 45 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQORiHkMTfRYlMHlF8s:-KQS-6Vh6_Z184u-YTHu:b-t23fko) ([raw file](https://github.com/robotlocomotion/drake/blob/3fab7cc6b437005272facdfc1a2cf78cfd3bcda5/drake/systems/framework/primitives/pid_controller.h#L45)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, is it clear what you would want this to do? Would `SetDefaultState()` restore a Context to the condition it was in right after `CreateDefaultContext()`? Or affect only continuous states? Other states are: difference states xd, modes, input and output port samples. Time and parameter values might also be included. Is this a proposal for a new family of System methods?
</details>
Not yet unless I see myself having to implement a `SetDefaultState` quite regularly. For now this method does what the documentation says for this particular system.

---

*[drake/systems/framework/primitives/pid_controller.h, line 22 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQRkNAnKCwo0J8ZJuEw:-KQRzyLlV9QlHLma-UTs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ed01c073c4646c1e80457eb877becb75f9c7999/drake/systems/framework/primitives/pid_controller.h#L22)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This comment needs updating since the same thing happens with the derivative constant (for a different input port).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r4.
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-:-KQS5vw3AXo2iDWD5ikk:bnfp4nl)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[drake/systems/framework/primitives/pid_controller.cc, line 63 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3275#-KQRj8Mpu1-Xz8btwTsO:-KQVuUnk2fZM5_9ZtM0q:b-bevt9h) ([raw file](https://github.com/robotlocomotion/drake/blob/1ed01c073c4646c1e80457eb877becb75f9c7999/drake/systems/framework/primitives/pid_controller.cc#L63)):*
> BTW, this could cause trouble if Kp and Kd can be changed after construction

I agree that would be trouble, and I think it should be prohibited, as it is on this Diagram.  There should be no such thing as "re-Finalize()", with or without Parameters.  However, it's a great point that sparsity is a function of Parameters.

> @david-german-tri, does the DiagramBuilder idea preclude this?

I think the DiagramBuilder helps communicate that this is disallowed: once you build a Diagram, it is fully baked and cannot be twiddled.  If you want different behavior, build a new Diagram.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3275)*
<!-- Sent from Reviewable.io -->
